server/coordinator/procedure/storage_impl.go

```diff
@@ -49,31 +53,66 @@ func NewEtcdStorageImpl(client *clientv3.Client, rootPath string, clusterID uint
 }
 
 // CreateOrUpdate example:
-// /{rootPath}/v1/procedure/{procedureID} -> {procedureType} + {procedureState} + {data}
-func (e EtcdStorageImpl) CreateOrUpdate(ctx context.Context, meta Meta) error {
+// /{rootPath}/v1/procedure/{procedureType}/{procedureID} ->  {procedureState} + {data}
+// ttl is only valid when greater than 0, if it is less than or equal to 0, it will be ignored.
+func (e EtcdStorageImpl) CreateOrUpdate(ctx context.Context, meta Meta, ttl int64) error {
 	s, err := encode(&meta)
 	if err != nil {
 		return errors.WithMessage(err, "encode meta failed")
 	}
-	keyPath := e.generaNormalKeyPath(meta.ID)
-	opPut := clientv3.OpPut(keyPath, s)
+
+	keyPath := e.generaNormalKeyPath(meta.Typ, meta.ID)
+
+	var opPut clientv3.Op
+	if ttl <= 0 {
+		opPut = clientv3.OpPut(keyPath, s)
+	} else {
+		// TODO: This implementation will cause each procedure to correspond to an etcd lease, which may cause too much pressure on etcd? May need to optimize implementation.
+		resp, err := e.client.Grant(ctx, ttl)
+		if err != nil {
+			return errors.WithMessage(err, "etcd get lease failed")
+		}
+		opPut = clientv3.OpPut(keyPath, s, clientv3.WithLease(resp.ID))
+	}
 
 	if _, err = e.client.Do(ctx, opPut); err != nil {
 		return errors.WithMessage(err, "etcd put data failed")
 	}
+
+	return nil
+}
+
+// Delete will delete the specified procedure,
+// and try to delete its corresponding history procedure if it exists.
+func (e EtcdStorageImpl) Delete(ctx context.Context, procedureType Typ, id uint64) error {
+	keyPath := e.generaNormalKeyPath(procedureType, id)
+	opDelete := clientv3.OpDelete(keyPath)
+
+	if _, err := e.client.Txn(ctx).Then(opDelete).Commit(); err != nil {
+		return err
+	}
+
+	deletedKeyPath := e.generaDeletedKeyPath(procedureType, id)
+	opDeleteMark := clientv3.OpDelete(deletedKeyPath)
+	// Try to delete history procedure if it exists.
+	keyExists := clientv3util.KeyExists(deletedKeyPath)
+	if _, err := e.client.Txn(ctx).If(keyExists).Then(opDeleteMark).Commit(); err != nil {
+		log.Warn("drop history procedure failed", zap.String("deletedKeyPath", deletedKeyPath), zap.Error(err))
+	}
+
 	return nil
 }
 
 // MarkDeleted Do a soft deletion, and the deleted key's format is:
 // /{rootPath}/v1/historyProcedure/{clusterID}/{procedureID}
-func (e EtcdStorageImpl) MarkDeleted(ctx context.Context, id uint64) error {
-	keyPath := e.generaNormalKeyPath(id)
+func (e EtcdStorageImpl) MarkDeleted(ctx context.Context, procedureType Typ, id uint64) error {
+	keyPath := e.generaNormalKeyPath(procedureType, id)
 	meta, err := etcdutil.Get(ctx, e.client, keyPath)
 	if err != nil {
 		return errors.WithMessage(err, "get meta failed")
 	}
 
-	deletedKeyPath := e.generaDeletedKeyPath(id)
+	deletedKeyPath := e.generaDeletedKeyPath(procedureType, id)
```

Comment:
Could you give an explanation about why the type of procedure is part of the key? It looks very weird.

Reply:
Considering query performance, currently we can only query all procedures through `List`. Using `Type` as part of the key can effectively reduce the amount of data queried in `list`.